Python UDFs (#8116)
fishtown-analytics/docs.getdbt.com: website/docs/docs/build/udfs.md

dbt creates, updates, and renames UDFs as part of DAG execution. The UDF is built in the warehouse before the model that references it. Refer to [listing and selecting UDFs](/docs/build/udfs#listing-and-selecting-udfs) for more info on how to build UDFs in your project.

> Refer to [Function properties](/reference/function-properties) or [Function configurations](/reference/function-configs) for more information on the configs/properties for UDFs.

Reviewer: added link to configs

## Defining UDFs in dbt

> You can define SQL and Python UDFs in dbt. Follow these steps to define UDFs in dbt:

Reviewer 2: ```suggestion
You can define SQL and Python UDFs in dbt. Note: Python UDFs are currently supported in Snowflake and BigQuery.

Follow these steps to define UDFs in dbt:
```

> 1. Create a SQL or Python file under the `functions` directory. For example:

Reviewer 2: ```suggestion
1. Create a SQL or Python file under the `functions` directory. For example, this UDF checks if a string represents a positive integer:
```

    <Tabs>

    <TabItem value="SQL">

>     Define a SQL UDF in a SQL file. For example, this UDF checks if a string represents a positive integer:

Reviewer 2: ```suggestion
    Define a SQL UDF in a SQL file.
```

    **Note**: You can specify configs in the SQL file or in the corresponding YAML file in Step 2. 

>     You can also create Python UDFs in the `functions` directory. For more information, see [Defining Python UDFs](#defining-python-udfs).

Reviewer 2: i think we should make this "Defining UDFs in dbt" section agnostic to whether it's a sql or python UDF

Reviewer: addressed https://github.com/dbt-labs/docs.getdbt.com/pull/8116/commits/d698658012b8c44b1ea289197a76aee49a90ffdb

    </File>

    </TabItem>
    <TabItem value="Python">

>     You can define Python UDFs in your dbt project and use them in SQL queries, both inside and outside of dbt. Python UDFs are currently supported in Snowflake and BigQuery.

Reviewer 2: ```suggestion
    Define a Python UDF in a Python file.
```

>     A Python UDF creates a Python function directly within your data warehouse, which you can invoke using SQL. This makes it easier to apply complex data transformations, calculations, or logic that would be difficult to define in SQL.

Reviewer 2: ```suggestion
```

Reviewer 2: this can just be covered in the FAQ i think

    <File name='functions/is_positive_int.py'>

    ```py
    import re
    
    def main(a_string):
        """

>         Return 1 if a_string represents a positive integer (e.g., "10", "+8" not allowed here),

Reviewer 2: is this an in-line description? should we put it in the yml instead and just have the python logic here

    </File>

    </TabItem>
    <TabItem value="Python">

>     Define a Python UDF in a Python file. The Python function returns 1 if `a_string` represents a positive integer (like "10", "+8" not allowed).

Reviewer 2: ```suggestion
    Define a Python UDF in a Python file. 
```

duplicative of what's in `description`, we don't have this for the sql example

    <File name='functions/is_positive_int.py'>

    ```py
    import re
    
    def main(a_string):
        return 1 if re.search(r'^[0-9]+$', a_string or '') else 0
    ```
    </File>
    </TabItem>
    </Tabs>

    **Note**: You can specify configs in the SQL file or in the corresponding YAML file in next step (Step 2). 

2. Specify the function name and define the config, properties, return type, and optional arguments in a corresponding YAML file. For example:

    <Tabs>
    <TabItem value="SQL">

>     <File name='functions/schema.yml'>

Reviewer 2: can we make `description` the same for the SQL version

> My UDF that returns 1 if a string represents a naked positive integer (like "10", "+8" is not allowed). # optional

    <TabItem value="Postgres">

    ```sql

>     CREATE OR REPLACE FUNCTION udf_schema.is_positive_in(a_string text)

Reviewer 2: ```suggestion
    CREATE OR REPLACE FUNCTION udf_schema.is_positive_int(a_string text)
```

      - [Snowflake](https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-introduction): `3.10`, `3.11`, `3.12`, and `3.13`
      - [BigQuery](https://cloud.google.com/bigquery/docs/user-defined-functions-python): `3.11`
    - [`entry_point`](/reference/resource-configs/entry-point) &mdash; Specify the Python function to be called.

    
    For example:

    <File name='functions/schema.yml'>

    ```yml
      functions:
        - name: is_positive_int
          description: My UDF that determines if a string represents a positive (+) integer. Returns 1 if a_string represents a positive integer (like "10", "+8" not allowed here). # optional
          config:
            runtime_version: "3.11"   # required
            entry_point: main      
            schema: udf_schema
            database: udf_db
          arguments:
            - name: a_string
              data_type: string
          returns:
            data_type: integer
    ```
    </File>

    This example generates the following `CREATE` UDF statement:

    <Tabs>

    <TabItem value="Snowflake">
    ```sql

>     CREATE OR REPLACE FUNCTION is_positive_int(a_string STRING)

Reviewer 2: ```suggestion
    CREATE OR REPLACE FUNCTION udf_db.udf_schema.is_positive_int(a_string STRING)
```

      RETURNS INTEGER
      LANGUAGE PYTHON
      RUNTIME_VERSION = '3.11'
      HANDLER = 'main'
    AS $$
    import re
    def main(a_string):
      return 1 if re.search(r'^[0-9]+$', a_string or '') else 0
    $$;
    ```
    </TabItem>

    <TabItem value="BigQuery">
    ```sql

>     CREATE OR REPLACE FUNCTION is_positive_int(a_string STRING)

Reviewer 2: ```suggestion
    CREATE OR REPLACE FUNCTION udf_db.udf_schema.is_positive_int(a_string STRING)
```

    ```
    </TabItem>
  
    </Tabs>
    </TabItem>
    <TabItem value="Python">
    
    The following configs are required when defining a Python UDF: 

    - [`runtime_version`](/reference/resource-configs/runtime-version) &mdash; Specify the Python version to run. Supported values are:
      - [Snowflake](https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-introduction): `3.10`, `3.11`, `3.12`, and `3.13`
      - [BigQuery](https://cloud.google.com/bigquery/docs/user-defined-functions-python): `3.11`
    - [`entry_point`](/reference/resource-configs/entry-point) &mdash; Specify the Python function to be called.

    
    For example:

    <File name='functions/schema.yml'>

    ```yml
      functions:
        - name: is_positive_int

>           description: My UDF that determines if a string represents a positive (+) integer. Returns 1 if a_string represents a positive integer (like "10", "+8" not allowed here). # optional

Reviewer 2: Ok just checking my understanding here...

The python logic returns a **1** if `^[0-9]+$` can be found in the string.

The sql logic returns the **starting position of** `^[0-9]+$`. Though since this is "any integer that's the first character of the string through the last" this would _have to be 1_ if it's found.

Ok, so I think the logic is indeed the same.

```suggestion
          description: My UDF that returns 1 if a string represents a naked positive integer (like "10", "+8" is not allowed). # optional
```

          config:
            runtime_version: "3.11"   # required
            entry_point: main      
            schema: udf_schema
            database: udf_db

>             volatility: deterministic  

Reviewer 2: the sql example has more callouts of what's #required vs. #7768 

can we add those into the python example too?

          arguments:
            - name: a_string

>               data_type: string

Reviewer 2: ```suggestion
              data_type: string
              description: The string that I want to check if it's representing a positive integer (like "10") 
```

making this same as sql example

          returns:
            data_type: integer
    ```
    </File>

>     This `functions/schema.yml` file example generates the following `CREATE` UDF statement:

Reviewer 2: I think this should actually be it's own step, not combined with #2.

It's _both_ the schema.yml file _and_ the sql/py file that generated the create statement **when you execute a `dbt build`** for this UDF.

`dbt build --select "resource_type:function"`
`dbt build --select is_positive_int`

Reviewer: ohh this is super important to clarify!

    <Tabs>

    <TabItem value="Snowflake">
    ```sql
    CREATE OR REPLACE FUNCTION udf_db.udf_schema.is_positive_int(a_string STRING)
      RETURNS INTEGER
      LANGUAGE PYTHON
      RUNTIME_VERSION = '3.11'
      HANDLER = 'main'
    AS $$
    import re
    def main(a_string):
      return 1 if re.search(r'^[0-9]+$', a_string or '') else 0
    $$;
    ```
    </TabItem>

    <TabItem value="BigQuery">
    ```sql
    CREATE OR REPLACE FUNCTION udf_db.udf_schema.is_positive_int(a_string STRING)
    RETURNS INT64
    LANGUAGE python
    OPTIONS(runtime_version="python-3.11", entry_point="main")
    AS r'''
      import re
      def main(a_string):
        return 1 if re.search(r'^[0-9]+$', a_string or '') else 0
    ''';
    ```
    </TabItem>
    </Tabs>
    </TabItem>
    </Tabs>

	:::info volatility warehouse-specific

>    	Something to note is that `volatility` is accepted in dbt for both SQL and Python UDFs, but the handling of it is warehouse-specific. BigQuery ignores `volatility` and dbt displays a warning. In Snowflake, `volatility` is applied when creating the UDF. Refer to [volatility](/reference/resource-configs/volatility#warehouse-specific-volatility-keywords) for more information.

Reviewer: ```suggestion
   	Something to note is that `volatility` is accepted in dbt for both SQL and Python UDFs, but the handling of it is warehouse-specific. BigQuery ignores `volatility` and dbt displays a warning. In Snowflake, `volatility` is applied when creating the UDF. Refer to [setting volatility](#setting-volatility-in-udfs) for more information.
```

      - [Snowflake](https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-introduction): `3.10`, `3.11`, `3.12`, and `3.13`
      - [BigQuery](https://cloud.google.com/bigquery/docs/user-defined-functions-python): `3.11`
    - [`entry_point`](/reference/resource-configs/entry-point) &mdash; Specify the Python function to be called.

    
    For example:

    <File name='functions/schema.yml'>

    ```yml
      functions:
        - name: is_positive_int
          description: My UDF that determines if a string represents a positive (+) integer. Returns 1 if a_string represents a positive integer (like "10", "+8" not allowed here). # optional
          config:
            runtime_version: "3.11"   # required
            entry_point: main      
            schema: udf_schema
            database: udf_db
            volatility: deterministic  # behavior is warehouse-specific; see note after this example
          arguments:
            - name: a_string
              data_type: string
          returns:
            data_type: integer
    ```
    </File>

>     Something to note is that `volatility` is accepted in dbt for Python UDFs, but the handling of it is warehouse-specific. BigQuery ignores `volatility` and dbt displays a warning. In Snowflake, `volatility` is applied when creating the UDF. Refer to [volatility](/reference/resource-configs/volatility) for more information.

Reviewer: ```suggestion
   :::info volatility warehouse-specific
   Something to note is that `volatility` is accepted in dbt for Python UDFs, but the handling  of it is warehouse-specific. BigQuery ignores `volatility` and dbt displays a warning. In Snowflake, `volatility` is applied when creating the UDF. Refer to [volatility](/reference/resource-configs/volatility#warehouse-specific-volatility-keywords) for more information.
    :::
```

    </TabItem>
  
    </Tabs>
    </TabItem>
    <TabItem value="Python">
    
    The following configs are required when defining a Python UDF: 

    - [`runtime_version`](/reference/resource-configs/runtime-version) &mdash; Specify the Python version to run. Supported values are:
      - [Snowflake](https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-introduction): `3.10`, `3.11`, `3.12`, and `3.13`
      - [BigQuery](https://cloud.google.com/bigquery/docs/user-defined-functions-python): `3.11`
    - [`entry_point`](/reference/resource-configs/entry-point) &mdash; Specify the Python function to be called.

    
    For example:

    <File name='functions/schema.yml'>

    ```yml
      functions:
        - name: is_positive_int

>           description: Returns 1 if a_string matches ^[0-9]+$, else 0

Reviewer 2: ```suggestion
          description: My UDF that determines if a string represents a positive (+) integer # optional
```

>           config:

Reviewer 2: I would add the same configs from the SQL example as well (`schema`, `database`, etc.)

@QMalcolm does python UDFs support volatility?

Reviewer 3: Depends on how you look at it 😅 
You can set a volatility on a UDF, and we shouldn't raise an error no matter what. What happens with that volatility though depends on the data warehouse
* For bigquery the volatility will be ignored + emit a warning (for both SQL and Python)
* For snowflake the volatility will be set for the creation of the UDF (for both SQL and Python)

Reviewer 3: In most data warehouses, it appears it's safe to assume (for the most part 🤞🏻) that volatility is supported the same across a given data warehouse's different UDF languages

    def secret_int_a():
      return 2

    def secret_int_b():
      return 3
    ```
    </File>

2. Specify the function name and define the configs, properties, return type, and optional arguments in a corresponding YAML file.

    The following configs are required when defining a Python UDF: 

    - `runtime_version` &mdash; Specify the Python version to run.

      <Expandable alt_header="Supported values">

      - [Snowflake](https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-introduction): `3.10`, `3.11`, `3.12`, and `3.13`
      - [BigQuery](https://cloud.google.com/bigquery/docs/user-defined-functions-python): `3.11`

      </Expandable>

    - `entry_point` &mdash; Specify the Python function to be called.

    For example:

    <File name='functions/schema.yml'>

    ```yml
    functions:
      - name: my_function
        description: Adds two secret numbers to initial number
        config:
          runtime_version: "3.11" # required
          entry_point: main # required
        arguments:
          - name: an_int
            data_type: int
        returns:
          data_type: int
    ```
    </File>

>     This example generates the following `CREATE` UDF statement in Snowflake:

Author: Do we have a sample statement for BigQuery?

Reviewer 3: The sample statement for bigquery would be
```sql
CREATE OR REPLACE FUNCTION my_function(an_int INT64)
RETURNS INT64
LANGUAGE python
OPTIONS(runtime_version="python-3.11", entry_point="main")
AS r'''
    def main(an_int):
      return an_int + secret_int_a() + secret_int_b()

    def secret_int_a():
      return 2

    def secret_int_b():
      return 3
''';
```

<Volatility />

### Defining Python UDFs

You can define Python UDFs in your dbt project and use them in SQL queries, both inside and outside of dbt. Python UDFs are currently supported in Snowflake and BigQuery. 

> A Python UDF creates a Python function directly within your data warehouse, which you can invoke using SQL. This makes it easier to apply complex data transformations, calculations, or logic that would be difficult to define in SQL.

Reviewer 2: we should add this reasoning to the FAQ on when to use a UDF instead of a macro -> https://docs.getdbt.com/faqs/Project/udfs-vs-macros

Reviewer: addressed https://github.com/dbt-labs/docs.getdbt.com/pull/8116/commits/d698658012b8c44b1ea289197a76aee49a90ffdb

    ```
    </File>

    In your DAG, a UDF node is created from the SQL/Python and YAML definitions, and there will be a dependency between `is_positive_int` → `my_model`.
   <Lightbox src="/img/docs/building-a-dbt-project/UDF-DAG.png" width="85%" title="The DAG for the UDF node" />

> After defining a UDF, if you update the SQL/Python file that contains its function body (`is_positive_int.sql` in this example) or its configurations, your changes will be applied to the UDF in the warehouse next time you `build`.

Reviewer 2: ```suggestion
After defining a UDF, if you update the SQL/Python file that contains its function body (`is_positive_int.sql` or `is_positive_int.py` in this example) or its configurations, your changes will be applied to the UDF in the warehouse next time you `build`.
```

    ```
    </File>

    In your DAG, a UDF node is created from the SQL/Python and YAML definitions, and there will be a dependency between `is_positive_int` → `my_model`.
   <Lightbox src="/img/docs/building-a-dbt-project/UDF-DAG.png" width="85%" title="The DAG for the UDF node" />

After defining a UDF, if you update the SQL/Python file that contains its function body (`is_positive_int.sql` or `is_positive_int.py` in this example) or its configurations, your changes will be applied to the UDF in the warehouse next time you `build`.

> ### Setting `volatility` in UDFs

Reviewer 2: I don't think this should be a section on this page. We don't talk in detail about any of the other configs. I think this should just be covered in the `volatility` config page.

> For more information about selecting UDFs, see the examples in [Node selector methods](/reference/node-selection/methods#file).

Reviewer 2: I think this section should be called "Listing and building UDFs" instead of "selecting"

## Limitations
- Creating UDFs in other languages (for example, Java or Scala) is not yet supported. 

> - Only <Term id="scalar">scalar</Term> functions are currently supported.

Reviewer 2: Let's also callout the limited warehouses that python UDFs are currently supported in here.
